File: bridge/ports.go
package main

// Map a dev-server port to the directory it was launched from — the first
// half of the routing cascade — and list every listening port for the setup
// page.
//
// Linux reads procfs directly instead of shelling out: `ss` needs iproute2 and
// a PATH, and a daemon started by a service manager gets a minimal
// environment. macOS has no procfs, so it asks lsof.

import (
	"bufio"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
)

type PortOwner struct {
	Pid int `json:"pid"`
	// Empty when the cwd is unreadable — another user, or ptrace_scope.
	Cwd string `json:"cwd"`
}

type ListeningPort struct {
	Port int
	Pid  int
	Cwd  string
}

func portStrategy() string {
	switch runtime.GOOS {
	case "linux":
		return "procfs"
	case "darwin":
		return "lsof"
	}
	return "none"
}

// Listening socket state in /proc/net/tcp.
const tcpListen = "0A"

var procNetTables = []string{"/proc/net/tcp", "/proc/net/tcp6"}

func parsePort(raw string) int {
	port, err := strconv.Atoi(raw)
	if err != nil || port <= 0 || port > 65535 {
		return 0
	}
	return port
}

// procListening maps each listening socket inode to its port. The bound
// address is ignored on purpose: a dev server can be on 0.0.0.0, 127.0.0.1,
// a LAN address or IPv6, and matching on the port alone holds across all.
func procListening() map[string]int {
	byInode := map[string]int{}
	for _, table := range procNetTables {
		file, err := os.Open(table)
		if err != nil {
			continue // no IPv6 on this kernel, or a container without procfs
		}
		scanner := bufio.NewScanner(file)
		scanner.Scan() // header
		for scanner.Scan() {
			fields := strings.Fields(scanner.Text())
			if len(fields) < 10 || fields[3] != tcpListen {
				continue
			}
			local := fields[1]
			port, err := strconv.ParseInt(local[strings.LastIndex(local, ":")+1:], 16, 32)
			if err == nil {
				byInode[fields[9]] = int(port)
			}
		}
		file.Close()
	}
	return byInode
}

// pidsByInode walks /proc once, however many inodes are asked about — which
// is what makes listing every port as cheap as looking up one.
func pidsByInode(inodes map[string]int) map[string][]int {
	found := map[string][]int{}
	if len(inodes) == 0 {
		return found
	}
	entries, err := os.ReadDir("/proc")
	if err != nil {
		return found
	}
	for _, entry := range entries {
		pid, err := strconv.Atoi(entry.Name())
		if err != nil {
			continue
		}
		fdDir := "/proc/" + entry.Name() + "/fd"
		handles, err := os.ReadDir(fdDir)
		if err != nil {
			continue // someone else's process, or it exited mid-scan
		}
		seen := map[string]bool{}
		for _, handle := range handles {
			link, err := os.Readlink(fdDir + "/" + handle.Name())
			if err != nil || !strings.HasPrefix(link, "socket:[") {
				continue
			}
			inode := link[len("socket:[") : len(link)-1]
			if _, wanted := inodes[inode]; wanted && !seen[inode] {
				seen[inode] = true
				found[inode] = append(found[inode], pid)
			}
		}
	}
	return found
}

func procCwd(pid int) string {
	cwd, err := os.Readlink("/proc/" + strconv.Itoa(pid) + "/cwd")
	if err != nil {
		return ""
	}
	return cwd
}

func lsofCwd(pid int) string {
	out, err := exec.Command("lsof", "-a", "-p", strconv.Itoa(pid), "-d", "cwd", "-Fn").Output()
	if err != nil {
		return ""
	}
	for _, line := range strings.Split(string(out), "\n") {
		if strings.HasPrefix(line, "n") {
			return line[1:]
		}
	}
	return ""
}

// ownersForPort returns every process listening on port. Several is normal:
// SO_REUSEPORT workers, or a supervisor sharing a listening fd with its child
// (Vite, Next) — those share a cwd and collapse in cwdsForPort.
func ownersForPort(raw string) []PortOwner {
	port := parsePort(raw)
	if port == 0 {
		return nil
	}
	var owners []PortOwner
	switch portStrategy() {
	case "procfs":
		inodes := map[string]int{}
		for inode, p := range procListening() {
			if p == port {
				inodes[inode] = p
			}
		}
		pids := map[int]bool{}
		for _, list := range pidsByInode(inodes) {
			for _, pid := range list {
				pids[pid] = true
			}
		}
		for pid := range pids {
			owners = append(owners, PortOwner{Pid: pid, Cwd: procCwd(pid)})
		}
	case "lsof":
		out, err := exec.Command("lsof", "-nP", "-iTCP:"+strconv.Itoa(port), "-sTCP:LISTEN", "-t").Output()
		if err != nil {
			return nil // lsof exits non-zero when nothing matches
		}
		for _, line := range strings.Fields(string(out)) {
			if pid, err := strconv.Atoi(line); err == nil {
				owners = append(owners, PortOwner{Pid: pid, Cwd: lsofCwd(pid)})
			}
		}
	}
	return owners
}

// cwdsForPort returns the distinct working directories behind a port, minus
// excludePid — the bridge passes its own, since a port it serves (its
// proxies) says nothing about any project. The whole set, not one string:
// collapsing several owners to the first is how a guess used to get made
// before routing ever saw the evidence.
func cwdsForPort(port string, excludePid int) []string {
	seen := map[string]bool{}
	var dirs []string
	for _, owner := range ownersForPort(port) {
		if owner.Pid == excludePid || owner.Cwd == "" || seen[owner.Cwd] {
			continue
		}
		seen[owner.Cwd] = true
		dirs = append(dirs, owner.Cwd)
	}
	return dirs
}

// listeningPorts returns every listening TCP port outside the ephemeral range.
// That range is where port-0 binds land — Next's internal router workers,
// debuggers, language servers — and a dev server someone opens is never there.
func listeningPorts() []ListeningPort {
	low, high := ephemeralRange()
	var all []ListeningPort
	switch portStrategy() {
	case "procfs":
		byInode := procListening()
		seen := map[string]bool{}
		for inode, pids := range pidsByInode(byInode) {
			for _, pid := range pids {
				// One port on IPv4 and IPv6 is two inodes held by one process.
				key := strconv.Itoa(byInode[inode]) + ":" + strconv.Itoa(pid)
				if seen[key] {
					continue
				}
				seen[key] = true
				all = append(all, ListeningPort{Port: byInode[inode], Pid: pid, Cwd: procCwd(pid)})
			}
		}
	case "lsof":
		all = lsofListening()
	}
	var result []ListeningPort
	for _, entry := range all {
		if entry.Port < low || entry.Port > high {
			result = append(result, entry)
		}
	}
	return result
}

func lsofListening() []ListeningPort {
	out, err := exec.Command("lsof", "-nP", "-iTCP", "-sTCP:LISTEN", "-Fpn").Output()
	if err != nil {
		return nil
	}
	seen := map[string]bool{}
	cwds := map[int]string{}
	var result []ListeningPort
	pid := 0
	for _, line := range strings.Split(string(out), "\n") {
		switch {
		case strings.HasPrefix(line, "p"):
			pid, _ = strconv.Atoi(line[1:])
		case strings.HasPrefix(line, "n") && pid != 0:
			port, err := strconv.Atoi(line[strings.LastIndex(line, ":")+1:])
			key := strconv.Itoa(port) + ":" + strconv.Itoa(pid)
			if err != nil || seen[key] {
				continue
			}
			seen[key] = true
			if _, ok := cwds[pid]; !ok {
				cwds[pid] = lsofCwd(pid)
			}
			result = append(result, ListeningPort{Port: port, Pid: pid, Cwd: cwds[pid]})
		}
	}
	return result
}

func ephemeralRange() (int, int) {
	if runtime.GOOS == "linux" {
		if raw, err := os.ReadFile("/proc/sys/net/ipv4/ip_local_port_range"); err == nil {
			fields := strings.Fields(string(raw))
			if len(fields) == 2 {
				low, errLow := strconv.Atoi(fields[0])
				high, errHigh := strconv.Atoi(fields[1])
				if errLow == nil && errHigh == nil {
					return low, high
				}
			}
		}
	}
	return 49152, 65535
}

File: src/ports.ts
import { execFile } from "node:child_process";
import { readdirSync, readFileSync, readlinkSync } from "node:fs";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

/** A process listening on the port, with its working directory when readable. */
export interface PortOwner {
  pid: number;
  /** Null when /proc/<pid>/cwd is unreadable — another user, or ptrace_scope. */
  cwd: string | null;
}

export type PortStrategy = "procfs" | "lsof" | "none";

/**
 * Map a dev-server port to the directory it was launched from — the first half
 * of the routing cascade.
 *
 * Linux reads procfs directly instead of shelling out to `ss`. Three reasons,
 * in increasing order of importance: `ss` needs iproute2 installed (it is, but
 * `lsof` was not, which is how the old code silently stopped routing on this
 * machine); its output needs a positional parse; and it needs a PATH. That last
 * one matters most — a daemon started by a service manager gets a minimal
 * environment, which is exactly why src/service.ts had to hardcode a PATH.
 * Reading files removes the last subprocess from the Linux path entirely.
 *
 * `ss -lntpH 'sport = :N'` stays the right command for a human debugging this
 * by hand; it just isn't shipped as a code path.
 */
export function portStrategy(): PortStrategy {
  if (process.platform === "linux") return "procfs";
  if (process.platform === "darwin") return "lsof";
  return "none";
}

/** Listening socket state in /proc/net/tcp. */
const TCP_LISTEN = "0A";
const PROC_NET_TABLES = ["/proc/net/tcp", "/proc/net/tcp6"];

function parsePort(raw: string): number | null {
  if (!/^\d+$/.test(raw)) return null;
  const port = Number.parseInt(raw, 10);
  return port > 0 && port <= 65_535 ? port : null;
}

/**
 * Inodes of the sockets listening on `port`.
 *
 * The bound address is deliberately ignored. A dev server can be on 0.0.0.0,
 * 127.0.0.1, a LAN address or a tailnet address, and in tcp6 the same port
 * appears with a 32-hex-digit address — matching on the port alone is the only
 * thing that holds across all of them.
 */
function listeningInodes(port: number): Set<string> {
  const target = `:${port.toString(16).toUpperCase().padStart(4, "0")}`;
  const inodes = new Set<string>();

  for (const table of PROC_NET_TABLES) {
    let contents: string;
    try {
      contents = readFileSync(table, "utf8");
    } catch {
      continue; // no IPv6 on this kernel, or a container without procfs
    }
    for (const line of contents.split("\n").slice(1)) {
      const fields = line.trim().split(/\s+/);
      const local = fields[1];
      const state = fields[3];
      const inode = fields[9];
      if (local === undefined || inode === undefined) continue;
      if (state !== TCP_LISTEN) continue;
      if (!local.endsWith(target)) continue;
      inodes.add(inode);
    }
  }
  return inodes;
}

/**
 * Which pids hold each of these socket inodes open — one walk of /proc however
 * many inodes are asked about, which is what makes listing every port as cheap
 * as looking up one.
 */
function pidsByInode(inodes: Set<string>): Map<string, number[]> {
  const wanted = new Map([...inodes].map((inode) => [`socket:[${inode}]`, inode]));
  const found = new Map<string, number[]>();

  let entries: string[];
  try {
    entries = readdirSync("/proc");
  } catch {
    return found;
  }

  for (const entry of entries) {
    if (!/^\d+$/.test(entry)) continue;
    let handles: string[];
    try {
      handles = readdirSync(`/proc/${entry}/fd`);
    } catch {
      continue; // someone else's process, or it exited mid-scan
    }
    const pid = Number.parseInt(entry, 10);
    for (const handle of handles) {
      let link: string;
      try {
        link = readlinkSync(`/proc/${entry}/fd/${handle}`);
      } catch {
        continue;
      }
      const inode = wanted.get(link);
      if (inode === undefined) continue;
      const pids = found.get(inode) ?? [];
      if (!pids.includes(pid)) pids.push(pid);
      found.set(inode, pids);
    }
  }
  return found;
}

/** Pids holding any of these socket inodes open. */
function pidsForInodes(inodes: Set<string>): number[] {
  return [...new Set([...pidsByInode(inodes).values()].flat())];
}

function cwdOf(pid: number): string | null {
  try {
    return readlinkSync(`/proc/${pid}/cwd`);
  } catch {
    return null;
  }
}

function procfsOwners(port: number): PortOwner[] {
  const inodes = listeningInodes(port);
  // Skip the /proc walk entirely when nothing is listening — the common case.
  if (inodes.size === 0) return [];

  return pidsForInodes(inodes).map((pid) => ({ pid, cwd: cwdOf(pid) }) satisfies PortOwner);
}

async function lsofCwd(pid: number): Promise<string | null> {
  try {
    const { stdout } = await execFileAsync("lsof", ["-a", "-p", String(pid), "-d", "cwd", "-Fn"]);
    const line = stdout.split("\n").find((entry) => entry.startsWith("n"));
    return line === undefined ? null : line.slice(1);
  } catch {
    return null;
  }
}

async function lsofOwners(port: number): Promise<PortOwner[]> {
  let pidOut: string;
  try {
    ({ stdout: pidOut } = await execFileAsync("lsof", ["-nP", `-iTCP:${port}`, "-sTCP:LISTEN", "-t"]));
  } catch {
    return []; // lsof exits non-zero when nothing matches
  }

  const owners: PortOwner[] = [];
  for (const raw of pidOut.split("\n")) {
    const trimmed = raw.trim();
    if (!/^\d+$/.test(trimmed)) continue;
    const pid = Number.parseInt(trimmed, 10);
    owners.push({ pid, cwd: await lsofCwd(pid) });
  }
  return owners;
}

/**
 * Every process listening on `port`. Several is normal and not an error:
 * SO_REUSEPORT workers, or a supervisor sharing an inherited listening fd with
 * its child (Vite, Next) — those share a cwd and collapse in `cwdsForPort`.
 */
export async function ownersForPort(port: string): Promise<PortOwner[]> {
  const parsed = parsePort(port);
  if (parsed === null) return [];
  switch (portStrategy()) {
    case "procfs":
      return procfsOwners(parsed);
    case "lsof":
      return lsofOwners(parsed);
    case "none":
      return [];
  }
}

/**
 * Distinct working directories behind a port.
 *
 * `excludePid` drops one process from the evidence — the bridge passes its own,
 * since a port it serves (its proxies) says nothing about any project.
 *
 * Returns the whole set rather than one string on purpose: collapsing several
 * owners to `head -1`, as the old code did, is how a guess got made before the
 * routing layer ever saw the evidence.
 */
export async function cwdsForPort(port: string, excludePid: number | null = null): Promise<string[]> {
  const owners = await ownersForPort(port);
  const dirs = new Set<string>();
  for (const owner of owners) {
    if (owner.pid === excludePid) continue;
    if (owner.cwd !== null && owner.cwd.length > 0) dirs.add(owner.cwd);
  }
  return [...dirs];
}

/** A TCP port something is listening on, and the process doing it. */
export interface ListeningPort {
  port: number;
  pid: number;
  cwd: string | null;
}

/**
 * Every listening TCP port, outside the ephemeral range. The range is where
 * port-0 binds land — Next's internal router workers, debuggers, language
 * servers — and a dev server a person opens in a browser is never there.
 */
export async function listeningPorts(): Promise<ListeningPort[]> {
  const [low, high] = ephemeralRange();
  const all = portStrategy() === "procfs" ? procfsListening() : portStrategy() === "lsof" ? await lsofListening() : [];
  return all.filter((entry) => entry.port < low || entry.port > high);
}

function procfsListening(): ListeningPort[] {
  const portByInode = new Map<string, number>();
  for (const table of PROC_NET_TABLES) {
    let contents: string;
    try {
      contents = readFileSync(table, "utf8");
    } catch {
      continue;
    }
    for (const line of contents.split("\n").slice(1)) {
      const fields = line.trim().split(/\s+/);
      const local = fields[1];
      const inode = fields[9];
      if (local === undefined || inode === undefined || fields[3] !== TCP_LISTEN) continue;
      const hex = local.slice(local.lastIndexOf(":") + 1);
      portByInode.set(inode, Number.parseInt(hex, 16));
    }
  }
  if (portByInode.size === 0) return [];

  const seen = new Set<string>();
  const result: ListeningPort[] = [];
  for (const [inode, pids] of pidsByInode(new Set(portByInode.keys()))) {
    const port = portByInode.get(inode);
    if (port === undefined) continue;
    for (const pid of pids) {
      // One port on both IPv4 and IPv6 is two inodes held by one process.
      const key = `${port}:${pid}`;
      if (seen.has(key)) continue;
      seen.add(key);
      result.push({ port, pid, cwd: cwdOf(pid) });
    }
  }
  return result;
}

async function lsofListening(): Promise<ListeningPort[]> {
  let stdout: string;
  try {
    ({ stdout } = await execFileAsync("lsof", ["-nP", "-iTCP", "-sTCP:LISTEN", "-Fpn"]));
  } catch {
    return [];
  }
  const pairs = new Map<string, { port: number; pid: number }>();
  let pid: number | null = null;
  for (const line of stdout.split("\n")) {
    if (line.startsWith("p")) pid = Number.parseInt(line.slice(1), 10);
    else if (line.startsWith("n") && pid !== null) {
      const port = Number.parseInt(line.slice(line.lastIndexOf(":") + 1), 10);
      if (Number.isInteger(port)) pairs.set(`${port}:${pid}`, { port, pid });
    }
  }
  const cwds = new Map<number, string | null>();
  const result: ListeningPort[] = [];
  for (const { port, pid: owner } of pairs.values()) {
    if (!cwds.has(owner)) cwds.set(owner, await lsofCwd(owner));
    result.push({ port, pid: owner, cwd: cwds.get(owner) ?? null });
  }
  return result;
}

function ephemeralRange(): [number, number] {
  if (process.platform === "linux") {
    try {
      const [low, high] = readFileSync("/proc/sys/net/ipv4/ip_local_port_range", "utf8").trim().split(/\s+/).map(Number);
      if (low !== undefined && high !== undefined && Number.isInteger(low) && Number.isInteger(high)) return [low, high];
    } catch {
      // fall through to the IANA range
    }
  }
  return [49_152, 65_535];
}
